### PR TITLE
Story/csi 753 create vol from snapshot

### PR DESCRIPTION
**Pull request by igorgonibm**
create vol from snapshot

### CODE REVIEW

File: controller/array_action/array_mediator_interface.py [hunk 1]
```diff
@@ -179,6 +202,19 @@ def get_snapshot(self, snapshot_name):
...
+    @abstractmethod
+    def get_snapshot_by_id(self, snapshot_id):
+        """
+        This function return snapshot info about the snapshot.
+        Args:
+            snapshot_id : id of the snapshot in the storage system
+        Returns:
+           Snapshot
+        Raises:
+            SnapshotNameBelongsToVolumeError
```
Comment (ArbelNathan) [inline controller/array_action/array_mediator_interface.py:214]: but in other storage systems snapshot IS a volume, how can you tell?

File: controller/array_action/array_mediator_xiv.py [hunk 2]
```diff
@@ -158,6 +164,33 @@ def create_volume(self, name, size_in_bytes, capabilities, pool, volume_prefix="
...
             raise controller_errors.PermissionDeniedError("create vol : {0}".format(name))
 
+    def copy_volume_from_snapshot(self, name, src_snap_name, src_snap_capacity_in_bytes, min_vol_size_in_bytes):
+        logger.debug(
+            "Copy snapshot {0} data to volume {1}. Snapshot capacity {2}. Minimal requested volume capacity {3}".format(
+                name, src_snap_name, src_snap_capacity_in_bytes, min_vol_size_in_bytes))
+        try:
+            logger.debug("Formatting volume {0}".format(name))
+            self.client.cmd.vol_format(vol=name)
+            logger.debug("Copying Snapshot {0} data to volume {1}.".format(name, src_snap_name))
```
Comment (ArbelNathan) [inline controller/array_action/array_mediator_xiv.py:174]: is it the right order?

Comment (igorgonibm) [reply controller/array_action/array_mediator_xiv.py:174]: yes. First we format and then copy

File: controller/array_action/array_mediator_interface.py (hunk 1 above)
Comment (igorgonibm) [reply controller/array_action/array_mediator_interface.py:214]: In a9k too but volume/snapshot has field master_name which is volume name it is created from.
If this value is empty it is volume, if not - snapshot
I assume other storages also has a way to check. Anyway in this ticket i couldn't consider all other storages.

File: controller/controller_server/utils.py [hunk 3]
```diff
@@ -184,14 +207,23 @@ def validate_publish_volume_request(request):
...
+
+def get_snapshot_id_info(snapshot_id):
+    return _get_object_id_info(snapshot_id, config.OBJECT_TYPE_NAME_SNAPSHOT)
+
+
+def _get_object_id_info(full_object_id, object_type):
+    logger.debug("getting {0} info for id : {1}".format(object_type, full_object_id))
+    splitted_object_id = full_object_id.split(config.PARAMETERS_OBJECT_ID_DELIMITER)
+    if len(splitted_object_id) != 2:
+        # TODO - add other exception
```
[the remark this answers is not in the document]
Comment (igorgonibm) [reply controller/controller_server/utils.py:220]: This is fixed by you in delete snapshot which I'll push directly after this ticket. Adding ticket number will cause merge problems.

File: controller/array_action/array_mediator_interface.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (erantzabari) [reply controller/array_action/array_mediator_interface.py:214]: Lets keep for now the two methods with the two different exceptions code and open a followup ticket to union it.

File: controller/controller_server/csi_controller_server.py [hunk 4]
```diff
@@ -131,6 +140,75 @@ def CreateVolume(self, request, context):
             context.set_details('an internal exception occurred : {}'.format(ex))
             return csi_pb2.CreateVolumeResponse()
 
+    def _copy_volume_from_snapshot(self, vol, src_snapshot_id, min_vol_size, array_mediator):
```
[the remark this answers is not in the document]
Comment (erantzabari) [reply controller/controller_server/csi_controller_server.py:143]: 1. maybe copy_to_existing_volume_from_snapshot (we need to distinguish between the transaction overall and specific action inside the transaction)
2. Since the logic is not specific for storage, it can be remain in the controller

File: controller/controller_server/csi_controller_server.py [hunk 5]
```diff
@@ -131,6 +140,75 @@ def CreateVolume(self, request, context):
...
+            logger.error("Exception raised while copying snapshot data to volume")
+            logger.exception(ex)
+            try:
+                self._rollback_create_volume_from_snapshot(array_mediator, vol.id)
+            except Exception:
+                pass
+            raise ex
+        return vol
+
+    @retry(Exception, tries=5, delay=1)
```
[the remark this answers is not in the document]
Comment (erantzabari) [reply controller/controller_server/csi_controller_server.py:168]: Lets open a new ticket to build up mechanism when NOT to do rollback.
Currently we will do rollback on each failure to avoid mistake in the white-listed exception

File: controller/controller_server/csi_controller_server.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (igorgonibm) [reply controller/controller_server/csi_controller_server.py]: Renamed to copy_to_existing_volume_from_snapshot

File: controller/controller_server/csi_controller_server.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (igorgonibm) [reply controller/controller_server/csi_controller_server.py:170]: CSI-1505

File: controller/array_action/array_mediator_interface.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (igorgonibm) [reply controller/array_action/array_mediator_interface.py]: Opened CSI-1506

[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/array_action/array_mediator_interface.py]: maybe we are not talking about the same thing.
I mean this duplication:
```
def get_snapshot(self, snapshot_name):
        logger.debug("Get snapshot : {}".format(snapshot_name))
        try:
            cli_snapshot = self.client.cmd.vol_list(vol=snapshot_name).as_single_element
        except xcli_errors.IllegalNameForObjectError as ex:
            logger.exception(ex)
            raise controller_errors.IllegalObjectName(ex.status)
        if not cli_snapshot:
            return None
        if not cli_snapshot.master_name:
            raise controller_errors.SnapshotNameBelongsToVolumeError(cli_snapshot.name, self.endpoint)
        array_snapshot = self._generate_snapshot_response(cli_snapshot)
        return array_snapshot

    def get_snapshot_by_id(self, snapshot_id):
        try:
            cli_snapshot = self.client.cmd.vol_list(wwn=snapshot_id).as_single_element
        except xcli_errors.IllegalValueForArgumentError as ex:
            logger.exception(ex)
            raise controller_errors.IllegalObjectID(ex.status)
        if not cli_snapshot:
            return None
        if not cli_snapshot.master_name:
            raise controller_errors.SnapshotIdBelongsToVolumeError(snapshot_id, self.endpoint)
        return self._generate_snapshot_response(cli_snapshot)
```

File: controller/controller_server/csi_controller_server.py [hunk 6]
```diff
@@ -90,22 +94,27 @@ def CreateVolume(self, request, context):
...
                         "volume was not found. creating a new volume with parameters: {0}".format(request.parameters))
 
                     array_mediator.validate_supported_capabilities(capabilities)
                     vol = array_mediator.create_volume(volume_full_name, size, capabilities, pool, volume_prefix)
-
                 else:
                     logger.debug("volume found : {}".format(vol))
 
-                    if not (vol.capacity_bytes == request.capacity_range.required_bytes):
+                    if not src_snapshot_id and vol.capacity_bytes != request.capacity_range.required_bytes:
```
[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/controller_server/csi_controller_server.py:107]: I'm not sure what you mean.
The issue that I'm referring to is:
when I undo this change, all tests pass.

And I think is change is important, so it should be covered (i.e. indicatively fail when undone)

File: controller/array_action/config.py [hunk 7]
```diff
@@ -8,3 +8,6 @@
 
 # volume context
 CONTEXT_POOL = "pool"
+
+SRC_OBJECT_TYPE_VOLUME = 1
```
[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/array_action/config.py]: I don't think the solution is just to remove it (it's not the first or last time this happens).

please add(/open a ticket to add) "allow-global-unused-variables=no" to our pylint configuration.

File: controller/controller_server/csi_controller_server.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/controller_server/csi_controller_server.py]: Then what is the purpose of `ArrayMediatorAbstract`?
Is there an interest in keeping entire cross-storage logic in 1 class (/file)?

I don't think the location convention is theoretical, and I do see argument sparing as a plus, especially when you didn't provide any cons

File: controller/controller_server/csi_controller_server.py [hunk 8]
```diff
@@ -131,6 +140,75 @@ def CreateVolume(self, request, context):
...
+        except controller_errors.VolumeNotFoundError as ex:
+            logger.error("Volume not found while copying snapshot data to volume")
+            logger.exception(ex)
+            raise ex
+        except Exception as ex:
+            logger.error("Exception raised while copying snapshot data to volume")
+            logger.exception(ex)
+            try:
+                self._rollback_create_volume_from_snapshot(array_mediator, vol.id)
+            except Exception:
```
[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/controller_server/csi_controller_server.py]: if rollback fails, it would still also raise the original exception.
I think it is better than `logger.exception`

File: controller/controller_server/csi_controller_server.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/controller_server/csi_controller_server.py:170]: I don't think this is about which exceptions cause rollback, I think it's about which exceptions cause retry.

we have always performed retires only on connection exception, and I don't see why it should be different here.

File: controller/controller_server/csi_controller_server.py [hunk 9]
```diff
@@ -131,6 +140,75 @@ def CreateVolume(self, request, context):
...
+                pass
+            raise ex
+        return vol
+
+    @retry(Exception, tries=5, delay=1)
+    def _rollback_create_volume_from_snapshot(self, array_mediator, vol_id):
```
[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/controller_server/csi_controller_server.py:171]: nevermind.
this should be decided as part of the retry exception unification ticket (CSI-1505)

File: controller/controller_server/csi_controller_server.py [hunk 10]
```diff
@@ -131,6 +140,75 @@ def CreateVolume(self, request, context):
...
+        logger.debug("Rollback copy volume from snapshot. Deleting volume {0}".format(vol_id))
+        array_mediator.delete_volume(vol_id)
+
+    def _handle_existing_vol_src_snap(self, vol, src_snapshot_id, context):
+        """
+        This function should create a snapshot from volume in the storage system.
```
[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/controller_server/csi_controller_server.py:177]: then could you please explain what does this docstring mean? so we could improve it

File: controller/controller_server/csi_controller_server.py [hunk 11]
```diff
@@ -131,6 +140,75 @@ def CreateVolume(self, request, context):
...
+        Args:
+            vol                : volume fetched or created in CreateVolume
+            src_snapshot_id    : id of snapshot we should copy to vol or None if volume shhuld not be copied
+            context            : CreateVolume response context
+        Returns:
+            if volume exists and is copy of specified snapshot - set context status to OK ad return CreateVolumeResponse
```
[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/controller_server/csi_controller_server.py]: please do not resolve this until you fix it *wherever* applicable (e.g. [here](https://github.com/IBM/ibm-block-csi-driver/blob/dd6407373031be9875366a8c8e5840af92ef2a62/controller/controller_server/csi_controller_server.py#L194))

File: controller/controller_server/csi_controller_server.py [hunk 12]
```diff
@@ -131,6 +140,75 @@ def CreateVolume(self, request, context):
...
+            return None
+        if vol_copy_src_object_id == src_snapshot_id:
+            logger.debug(
+                "Volume {0} exists and it is copy of Snapshot {1}.".format(vol_name, src_snapshot_id))
+            context.set_code(grpc.StatusCode.OK)
+            return csi_pb2.CreateVolumeResponse()
+        else:
+            logger.debug(
+                "Volume {0} exists but it is not copy of Snapshot {1}.".format(vol_name, src_snapshot_id))
+            context.set_details("Volume was already exists but created but from different source.")
```
[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/controller_server/csi_controller_server.py]: I don't think this requires a separate ticket, so please make sure you get to it before merge

File: controller/array_action/array_mediator_xiv.py [hunk 13]
```diff
@@ -197,6 +230,19 @@ def get_snapshot(self, snapshot_name):
         array_snapshot = self._generate_snapshot_response(cli_snapshot)
         return array_snapshot
 
+    def get_snapshot_by_id(self, snapshot_id):
+        cli_snapshot = None
+        try:
+            cli_snapshot = self.client.cmd.vol_list(wwn=snapshot_id).as_single_element
+        except xcli_errors.IllegalValueForArgumentError as ex:
```
[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/array_action/array_mediator_xiv.py:236]: let me rephrase:
why is this not needed in `get_volume_name`?

File: controller/controller_server/csi_controller_server.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/controller_server/csi_controller_server.py]: 1. why are the existing `un/map_volume_by_initiators` "ok", but `copy_to_existing_volume_from_snapshot` isn't? what makes it less "storage behavior"?
2. if existing `ArrayMediatorAbstract` is the wrong way to divide the code, then what is the right way? we can't just keep adding all cross-storage code to a single controller class.

File: controller/controller_server/csi_controller_server.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/controller_server/csi_controller_server.py]: I mean that in the code I suggested, `rollback_ex` will also [contain the original exception](https://www.geeksforgeeks.org/python-raising-an-exception-to-another-exception/) as an inner exception.
so no need for `logger.exception`

File: controller/controller_server/csi_controller_server.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/controller_server/csi_controller_server.py]: `"Volume already exists but it was created but from a different source."` doesn't sound like a valid sentence to me.
please consult with a 3rd person (e.g. rivka)

File: controller/array_action/array_mediator_interface.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/array_action/array_mediator_interface.py]: I see that the exceptions are different...
it doesn't mean there isn't a lot of unnecessary code duplication.
I think the duplication here is pretty obvious, even if it's not 100% of the function.

File: controller/array_action/config.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/array_action/config.py]: when we find a gap in review, if we know of an easy way to prevent it from re-happening in the future, we include it as part of the review.
the usual way is adding a unit test to cover it, but this time it's even simpler!
and no part of the project is exclusive to a single contributor, on the contrary - we should all acquire experience in all parts, especially in a general linter.

File: controller/tests/array_action/xiv/array_mediator_xiv_tests.py [hunk 14]
```diff
@@ -91,6 +91,61 @@ def test_create_volume__generate_volume_response_raise_exception(self, response)
         with self.assertRaises(Exception):
             self.mediator.create_volume("vol", 10, [], "pool1")
 
+    def test_copy_volume_from_snapshot_succeeds_with_resize(self):
```
Comment (oriyarde) [inline controller/tests/array_action/xiv/array_mediator_xiv_tests.py:94]: please align all tests function to be similar to the name of the function which they test

File: controller/controller_server/csi_controller_server.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/controller_server/csi_controller_server.py]: the code can even be simple as:
```
except Exception as ex:
    self._rollback_create_volume_from_snapshot(array_mediator, vol.id)
    raise ex
```

File: controller/controller_server/csi_controller_server.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/controller_server/csi_controller_server.py]: I wish to avoid dictating complex articulation.
I want you to write whatever you originally meant, just with valid grammar

File: controller/controller_server/csi_controller_server.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/controller_server/csi_controller_server.py:107]: please comment here before resolving, so reviewers can tell how/why was this resolved

File: controller/tests/array_action/xiv/array_mediator_xiv_tests.py (hunk 14 above)
Comment (oriyarde) [reply controller/tests/array_action/xiv/array_mediator_xiv_tests.py:94]: I don't see any change here.
if you think no change is needed, [please comment in the conversation before resolving](https://github.com/IBM/ibm-block-csi-driver/pull/158#discussion_r432949827), otherwise it would be harder to keep track.
I think this test function name should be changed for the same reasons that the tested function name was changed, and for naming consistency.

Comment (oriyarde) [reply controller/tests/array_action/xiv/array_mediator_xiv_tests.py]: it is only [reflected now](https://github.com/IBM/ibm-block-csi-driver/pull/158/commits/dd709c9821bba60e66d76b8679f7e6b367d25f38).
so just avoid resolving before ensuring it is reflected in the "files" tab diff

File: controller/controller_server/csi_controller_server.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/controller_server/csi_controller_server.py:107]: maybe I misunderstood the purpose of adding `not src_snapshot_id`.
if there is a capacity mismatch, shouldn't we return `ALREADY_EXISTS`, regardless of `src_snapshot_id`?

[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/controller_server/csi_controller_server.py:107]: then I think we are missing a unit test that makes sure `ALREADY_EXISTS` is not returned on "from snapshot" scenario

File: controller/controller_server/csi_controller_server.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/controller_server/csi_controller_server.py]: yes, but the exception from `if_rollback_create_volume_from_snapshot` will "pass through" this except clause and "pick up" the original exception "on the way" (implicitly)

File: controller/controller_server/csi_controller_server.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (igorgonibm) [reply controller/controller_server/csi_controller_server.py:107]: Will open general ticket for open issues

File: controller/controller_server/csi_controller_server.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (igorgonibm) [reply controller/controller_server/csi_controller_server.py]: done

**Comment by igorgonibm** [comment]
Handle open conversations ticket: CSI-1509

File: controller/controller_server/csi_controller_server.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/controller_server/csi_controller_server.py:166]: leaving unresolved for future ticket

File: controller/controller_server/csi_controller_server.py [hunk 15]
```diff
@@ -132,6 +141,70 @@ def CreateVolume(self, request, context):
...
+        if vol_copy_src_object_id == src_snapshot_id:
+            logger.debug(
+                "Volume {0} exists and it is a copy of snapshot {1}.".format(vol_name, src_snapshot_id))
+            context.set_code(grpc.StatusCode.OK)
+            return csi_pb2.CreateVolumeResponse()
+        else:
+            logger.debug(
+                "Volume {0} exists but it is not a copy of snapshot {1}.".format(vol_name, src_snapshot_id))
+            context.set_details("Volume already exists but it was created from a different source.")
+            context.set_code(grpc.StatusCode.INTERNAL)
```
Comment (oriyarde) [inline controller/controller_server/csi_controller_server.py:196]: according to the [spec](https://github.com/container-storage-interface/spec/blob/v1.2.0/spec.md#createvolume):
"If a volume corresponding to the specified volume `name` already exists, is accessible from `accessibility_requirements`, and is compatible with the specified `capacity_range`, `volume_capabilities` and `parameters` in the `CreateVolumeRequest`, the Plugin MUST reply `0 OK` with the corresponding `CreateVolumeResponse`."

and since `volume_content_source` is not one of name/capacity/capabilities/parameters, we should return `OK` (and not `INTERNAL`)

Comment (igorgonibm) [reply controller/controller_server/csi_controller_server.py:196]: I don't think OK is correct. OK means all created or already exists as required.
Maybe they just didn't consider snapshots when adding this description. 
If you think it needs to be changed open a separate ticket.

Comment (oriyarde) [reply controller/controller_server/csi_controller_server.py:196]: I think it would be better if we comply with the spec, instead of assuming the spec is wrong.
I opened ticket CSI-1540

File: controller/array_action/config.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply controller/array_action/config.py]: CSI-1660

File: controller/array_action/array_mediator_interface.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply controller/array_action/array_mediator_interface.py]: [done](https://github.com/IBM/ibm-block-csi-driver/blob/1411d1d02f4689d8f93d6c8fba8e558f89ac4485/controller/array_action/errors.py#L211)

[the remark this answers is not in the document]
Comment (ArbelNathan) [reply controller/array_action/array_mediator_interface.py]: and CSI-1506 is redundant

File: controller/controller_server/csi_controller_server.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply controller/controller_server/csi_controller_server.py:166]: CSI-1505

File: controller/controller_server/csi_controller_server.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply controller/controller_server/csi_controller_server.py]: https://github.com/IBM/ibm-block-csi-driver/pull/483

File: controller/controller_server/csi_controller_server.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply controller/controller_server/csi_controller_server.py:107]: CSI-4052
